lib/notifications.dart

```diff
@@ -0,0 +1,72 @@
+import 'package:flutter/foundation.dart';
+
+import 'log.dart';
+import 'model/binding.dart';
+
+class NotificationService {
+  static NotificationService get instance => (_instance ??= NotificationService._());
+  static NotificationService? _instance;
+
+  NotificationService._();
+
+  /// Reset the state of the [NotificationService], for testing.
+  ///
+  /// TODO refactor this better, perhaps unify with ZulipBinding
+  @visibleForTesting
+  static void debugReset() {
+    instance.token.dispose();
+    instance.token = ValueNotifier(null);
+  }
+
+  /// The FCM registration token for this install of the app.
+  ///
+  /// This is unique to the (app, device) pair, but not permanent.
+  /// Most often it's the same from one run of the app to the next,
+  /// but it can change either during a run or between them.
+  ///
+  /// See also:
+  ///  * Upstream docs on FCM registration tokens in general:
+  ///    https://firebase.google.com/docs/cloud-messaging/manage-tokens
+  ValueNotifier<String?> token = ValueNotifier(null);
+
+  Future<void> start() async {
+    if (defaultTargetPlatform != TargetPlatform.android) return; // TODO(#321)
+
+    await ZulipBinding.instance.firebaseInitializeApp();
+
+    // TODO(#324) defer notif setup if user not logged into any accounts
+    //   (in order to avoid calling for permissions)
+
+    ZulipBinding.instance.firebaseMessagingOnMessage.listen(_onRemoteMessage);
+
+    // Get the FCM registration token, now and upon changes.  See FCM API docs:
+    //   https://firebase.google.com/docs/cloud-messaging/android/client#sample-register
+    ZulipBinding.instance.firebaseMessaging.onTokenRefresh.listen(_onTokenRefresh);
+    await _getToken();
+  }
+
+  Future<void> _getToken() async {
+    final value = await ZulipBinding.instance.firebaseMessaging.getToken(); // TODO(log) if null
```

Comment:
Logging might be right for now, but I wonder about:

- retrying
- preparing to warn the user that setup failed or is not complete

I think this request will naturally fail in ordinary circumstances, like no Internet connection, no Google Play Services, etc., right?

I assume it might also hang indefinitely in some failure modes.

I think those could be TODOs for now though.

Reply:
Yeah, when it fails we'll want to note that for showing a warning (as in #323). Which, it's true, `TODO(log)` doesn't cover.



> I assume it might also hang indefinitely in some failure modes.

Yeah, could be. We'll want #323 or related warnings to appear for the user if we haven't gotten a result here.